services/venice-router/src/main/java/com/linkedin/venice/router/api/routing/helix/HelixGroupLeastLoadedStrategy.java

```diff
@@ -60,10 +61,21 @@ public int selectGroup(long requestId, int groupCount) {
       }
       for (int i = 0; i < groupCount; ++i) {
         int currentGroup = (i + startGroupId) % groupCount;
-        long currentGroupCounter = counters[currentGroup];
+        int currentGroupCounter = counters[currentGroup];
         if (currentGroupCounter < smallestCounter) {
           smallestCounter = currentGroupCounter;
           leastLoadedGroup = currentGroup;
+          lowestAvgLatency = helixGroupStats.getGroupResponseWaitingTimeAvg(currentGroup);
+        } else if (currentGroupCounter == smallestCounter) {
+          double currentGroupAvgLatency = helixGroupStats.getGroupResponseWaitingTimeAvg(currentGroup);
+          /**
+           * Here we don't check whether {@link #currentGroupAvgLatency} is less than 0 or not, as when the group is not
+           * being used at all, the average latency will be -1.0.
+           */
+          if (currentGroupAvgLatency < lowestAvgLatency) {
```

Comment:
I was wondering if we should do `<=` check to round-robin the load across groups?

Reply:
Latency is very subtle as it is `double` type, and the difference between different group won't be exactly same, so `<=` check won't make any difference.

Reply:
If the pending request counter is same among several different zones and latency is truly same, `startGroupId` will do the RR load balancing.